### DISCUSSION TITLE
Radical Red Save

### DISCUSSION THREAD

**Discussion by evanwporter** [andrewbenington/OpenHome#49] [General]
With ca39be4 OpenHome has been configured to work with Radical Red saves. There are still some kinks that need to be worked out, but it works well enough right now.

**Comment by evanwporter** [comment]
Tests:
![image](https://github.com/user-attachments/assets/aae65b32-5348-416e-a4d4-858fad971ab6)
* ✅ Correctly displays boxes
![image](https://github.com/user-attachments/assets/b21a36a2-23f6-485e-9836-4cd8ea416177)
* ✅ Able to move Pokemon within OpenHome (Ho-Oh is from OH while Nihilego is from RR)
![image](https://github.com/user-attachments/assets/b3e570a3-204b-41c5-8eb5-bc71119c1037)
* ✅ Moved Pokemon appear in RR game
* ✅ Checksums are correctly calculated
![image](https://github.com/user-attachments/assets/81a44443-81cf-4f0c-bdbd-24739f8d41cb)
![image](https://github.com/user-attachments/assets/5aeae4a4-904d-4977-be25-d63a351cb337)
* ✅ OT transfers
* ✅ Item transfers correctly
* ✅ Level Transfers
* ✅ Moves Transfer correctly
* ❌ Pokerus?? I'm not sure why Pokerus was added unless it was on the pokemon to start


Todo:
* Item transfers
* ~~Moves Transfer~~
* ~~Create image for RR (kinda)~~
* ~~Add better detection of RR saves~~
* Change RR name from Saphire to Radical Red

@andrewbenington

**Comment by evanwporter** [comment]
With support for Radical Red implemented at a basic level I wanted to list off a few things that I think are important to think about:
* ~~The displayed sprites should be changed to either the OHPKM version or preferably the actual in-game sprites ([here](https://github.com/funnotbun/funnotbun.github.io/tree/main/data/species/frontspr)). I think the second option is preferable here but may take some extra work since I'm not sure how to do this. It won't impact the final file size that much since all the images are probably less than a MB. Currently the sprites for PK3RR are a zoomed up version of the gen 9 sprites.~~ ~~Changed it to OHPKM sprites in 330f100.~~ Change it to PK3RR sprites in 2d2cc6e. The source location of the pokemon file has a disconnect between the number on the file and the pokedex number, so more work needs to be done here.
* Radical Red is not for everyone, so I think there should be an option to enable support for third party rom hacks (maybe even games made in RPG maker???). This option is off by default. Again I am unsure how to do this, since I haven't examined the renderer in depth. Once this is done I would feel good about merging the Radical Red branch into the main branch
* ~~The item seems to change when switching between PK3RR and other saves. Not sure why since I believe OH imports pokemon using the string. Maybe you have an idea. Also the items should use the general image~~ Fixed this in `pokemon-files-js`.
* The detection system for RR saves is shaky at best. Ideally there is a popup when it can't tell which game the save belongs to like in PKHex. Again not sure how to do this. Front end is not my forte.
* Fakemon. Currently Sevii forms (exclusive to Radical Red and only of some Pokemon (Lokix for example)) show up that in their non regional form. IE: Lokix-Sevii becomes Lokix. Thus the Sevii form is lost. Other than the Sevii forms I think the only fakemon is a pokemon from Palworld no clue how this is exposed to OpenHome. What would need to be done to incorporate these fakemon into OpenHome? Would there need to be a separate list from the National Dex? Also if a seperate list of Radical Red Pokemon is created then it would be cool to switch the stats and typing when PK3RR is selected as a possible file type ([Stats for every pokemon in RR](https://github.com/funnotbun/funnotbun.github.io/blob/main/data/species/Base_Stats.c)).
* In the future it should be very easy to incorporate pokemon from other Rom hacks, given a list of items, species, and moves and all of their corresponding hexadecimal values (ie: like [this](https://github.com/Skeli789/Dynamic-Pokemon-Expansion/blob/master/include/species.h)). But it should just be plug in and go. Of course a generalized subclass of PK3 would be good.
* Also this needs to be changed 
![image](https://github.com/user-attachments/assets/aa667fc7-e7a8-40e3-80f1-ad6674c1b00d)

**Reply by evanwporter** [reply]
@andrewbenington 

How difficult would the second bullet point be to implement? Ie: an `Experimental Rom Hacks Support` option

**Reply by andrewbenington** [reply]
I actually just started working on that. Unfortunately I didn't code a central source of truth for "all save file types", so that's my next goal. Once that's a thing it should be pretty easy to enable/disable different save type support

**Reply by andrewbenington** [reply]
Also for now I think we should just grey out any fakemons and "fakeformes" (sevii formes, etc), because those are never going to have "real" pokedex numbers. If you come up with a more elegant solution I'm open to it, but it's not going to be a simple task, especially as new official pokemon are dropped and take up new dex numbers

**Comment by andrewbenington** [comment]
Just pushed some updates to the radical-red branch. you might have to change `package.json` back to point `pokemon-files` at your local version again, sorry about that.

(here's the major commit: https://github.com/andrewbenington/OpenHome/commit/0bebb3ee32e78a617d9999de2c24bcf8e48edc10)

**Reply by andrewbenington** [reply]
@evanwporter Actually this is going to be a significant refactor so I'm pulling it into its own branch, that way it doesn't get in the way of your work ([optional-savetypes](https://github.com/andrewbenington/OpenHome/tree/optional-savetypes))

**Reply by evanwporter** [reply]
Correct me if I'm wrong but to add an option called `Enable Experimental Rom Hack Support` could you just store this boolean somewhere. Then whenever there is an option to save to `PK3RR` you add a `AND` condition with this boolean. Support for standard games should be always enabled/can't be turned off. Also `SAVtypes/util.ts` in the `getSaveType` function should be access this option.

Now afaik CFRU hacks are unique in that they are the only ones to significantly modify how the file is Saved. Most other gen 3 rom hacks only expand the pokedex which doesn't require modifications to the save since every already has enough space to represent an expanded pokedex. As for gen 4+ rom hacks, they don't change anything really about the save and can be read normally. (Hopefully this changes with Drayano's Aurora Crystal hack!)

With all this being said. I think centralizing the save types is a good idea. An example of where this could be very helpful is in Gen 5 where B2W2 have the poke studios. Or in Gen 4 where the pokeathalon records & leaves could be saved and displayed in OH. ~~Actually if we could display the pokeathalon records/methods for each pokemon...that would be really cool. It would add another level of collectibility to the ribbon master challenge.~~ Did some research and I don't the exact offsets have been found yet..

**Reply by andrewbenington** [reply]
Yeah i just think it's a good opportunity to have one place where all supported saves are defined, that a settings screen could alter to let people enable and disable specific ones. you're absolutely right that it would be easy to make that a toggle somewhere, but in the future we don't want to have hardcoded toggles for each rom hack save type anyways

**Reply by andrewbenington** [reply]
And yeah I never got around to detailed display of stuff like pokeathlon stats, size and weight, and pokerus status but that is eventually the goal

**Comment by evanwporter** [comment]
![image](https://github.com/user-attachments/assets/1524004c-dec6-48a4-b1e0-42110652405d)

Species images for RR will need to be updated. Like what even is this.

**Reply by evanwporter** [reply]
With c0d6ab4 the sprites have been fixed.

**Comment by andrewbenington** [comment]
@evanwporter Sorry for the huge changes that happened to the main branch, I tried my best to merge them into the `radical-red` branch while keeping all your code intact. I moved some logic to the G3RRSAV file, specifically save file detection (i.e. how do we tell this is a radical red save) and pokemon species/forme support.

I've done this for all the other save file types too. Now the save file detection process just goes through a list of supported file types and tests each one until the right format is found. "Extra" save file types like Radical Red are checked first because I assume those checks will be more specific.

The goal should be to keep the Radical Red code in as few files as possible so it can be eventually moved to its own plugin. This will probably require more big changes to make things more modular, but once that happens it should be much more straightforward to add support for other ROM hacks as well!

**Comment by andrewbenington** [comment]
@evanwporter Here's how I think Game of Origin should be handled for rom hacks like Radical Red:

Every plugin will have a string "identifier", like `radical_red`. On the radical red branch, I added a field to OHPKM to store this as a string of length up to 32, called `pluginOrigin`. Pokémon originating from a rom hack will have the vanilla game as its game of origin, and the ROM hack identifier as its plugin origin. For example, Radical Red mons with have a game of origin of `GameOfOrigin.FireRed`, and a `pluginOrigin` of `radical_red`. This way the GameOfOrigin enum stays reserved for official formats, while letting us track ROM hack origins as well.

**Comment by evanwporter** [comment]
![image](https://github.com/user-attachments/assets/93ba6e41-a8ef-4077-8596-ac178513938a)

Added an `isLocked` method to lock certain restricted pokemon (ie: fakemons/fakeforms) from being dragged/selected. For this I just the `disabled` property and then added another line to disable dragging for locked mons.

@andrewbenington I'm excited to say that RR has almost been fully integrated. Two things left:

(1) Dealing with fakemons (chillet from palworld) sprites <-- ~~I do not know how chillet might transfer~~ It should have a dexNum of -1. This may then turn up as None.
(2) Dealing with fake moves. Currently they are zeroed out, and I think this is the best way to handle it...unless you have a better idea.

**Comment by evanwporter** [comment]
Also the file select menu seems to have bugged out with your most recent commit. Displays `K3RR`. Not sure what happened here...

**Reply by evanwporter** [reply]
Unfortunately it does not seem to be building anymore due to the uintarray stuff.

**Reply by andrewbenington** [reply]
ugh i'm sorry, can you post the specific error? i'm busy tonight otherwise i'd fix it myself

javascript's APIs for working with raw bytes give me such a headache

**Reply by evanwporter** [reply]
Honestly there's a lot of issues.

```
src/types/pkm/OHPKM.ts:196:29 - error TS2339: Property 'isShadow' does not exist on type 'PKMInterface | PluginPKMInterface | OHPKM'.
  Property 'isShadow' does not exist on type 'PKMInterface'.

196       this.isShadow = other.isShadow ?? false
                                ~~~~~~~~

src/types/pkm/OHPKM.ts:301:7 - error TS2740: Type 'ArrayBuffer' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteOffset, copyWithin, and 29 more.

301       this.homeTracker = other.homeTracker ?? new Uint8Array(8)
          ~~~~~~~~~~~~~~~~

src/types/pkm/OHPKM.ts:312:9 - error TS2322: Type 'Uint8Array | ArrayBuffer' is not assignable to type 'Uint8Array'.
  Type 'ArrayBuffer' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteOffset, copyWithin, and 29 more.

312         this.trFlagsSwSh = other.trFlagsSwSh
            ~~~~~~~~~~~~~~~~

src/types/pkm/OHPKM.ts:315:9 - error TS2322: Type 'Uint8Array | ArrayBuffer' is not assignable to type 'Uint8Array'.
  Type 'ArrayBuffer' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteOffset, copyWithin, and 29 more.

315         this.tmFlagsBDSP = other.tmFlagsBDSP
            ~~~~~~~~~~~~~~~~

src/types/pkm/OHPKM.ts:327:9 - error TS2322: Type 'Uint8Array | ArrayBuffer' is not assignable to type 'Uint8Array'.
  Type 'ArrayBuffer' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteOffset, copyWithin, and 29 more.

327         this.moveFlagsLA = other.moveFlagsLA
            ~~~~~~~~~~~~~~~~

src/types/pkm/OHPKM.ts:330:9 - error TS2322: Type 'Uint8Array | ArrayBuffer' is not assignable to type 'Uint8Array'.
  Type 'ArrayBuffer' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteOffset, copyWithin, and 29 more.

330         this.tutorFlagsLA = other.tutorFlagsLA
            ~~~~~~~~~~~~~~~~~

src/types/pkm/OHPKM.ts:333:9 - error TS2322: Type 'Uint8Array | ArrayBuffer' is not assignable to type 'Uint8Array'.
  Type 'ArrayBuffer' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteOffset, copyWithin, and 29 more.

333         this.masterFlagsLA = other.masterFlagsLA
            ~~~~~~~~~~~~~~~~~~

src/types/pkm/OHPKM.ts:359:9 - error TS2322: Type 'Uint8Array | ArrayBuffer' is not assignable to type 'Uint8Array'.
  Type 'ArrayBuffer' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteOffset, copyWithin, and 29 more.

359         this.tmFlagsSV = other.tmFlagsSV
            ~~~~~~~~~~~~~~

src/types/pkm/OHPKM.ts:1667:7 - error TS2740: Type 'ArrayBuffer' is missing the following 
properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteOffset, copyWithin, and 
29 more.

1667       this.tmFlagsSV = other.tmFlagsSV
           ~~~~~~~~~~~~~~


Found 9 errors in the same file, starting at: src/types/pkm/OHPKM.ts:196
```

**Reply by andrewbenington** [reply]
can you check the version of `pokemon-files` in package.json? i think the most recent should be 0.3.5. i think these issues are because of an outdated version of that library

**Reply by evanwporter** [reply]
Yeah honestly I have no idea why the builder is so finnicky. But I have it built right now I can make updates and run it using the `run dev` which is ultimately I need. We can figure out the build issues later.

Regarding the `K3RR` issue I think it has to do with `PK3RR` not being a part of the `PKM`. So I created a `PluginPKM` to begin to address that issue.

**Reply by evanwporter** [reply]
Once we finish this I feel really good merging the radical red branch in OH in to main. The `pokemon-files` rr branch can be scrapped

**Reply by andrewbenington** [reply]
the K3RR thing is because of a shortcut i used to get the class name, i remove the first character which is usually an underline. i will investigate

**Reply by andrewbenington** [reply]
and yes i think so too. that way the item bag branch will only differ from the main branch with the item bag stuff. it was confusing having radical red code there that wasn't on the main branch

**Reply by andrewbenington** [reply]
@evanwporter I just pushed a fix to the `K3RR` issue. What I think happens is under the hood is Node renames classes from imported libraries to have their name prefixed with an underscore, which stopped happening for `PK3RR` when I moved it into this repo. So I changed the code to remove the first underscore in that file type select (before I was blindly removing the first character regardless)

**Comment by andrewbenington** [comment]
@evanwporter can we remove `src/renderer/public/sprites/rr/_rename.js` and the `src/renderer/public/sprites/rr/rr` directory? i assume they were used for generating the sprites, but aren't used in the final app

**Reply by evanwporter** [reply]
Can we move the `_rename.js` to the scripts/generate folder? The subdir `rr/rr` can be removed. I want to be able to use `_rename.js` again in case there is an issue with one of the sprites or something...

**Reply by andrewbenington** [reply]
yep sounds good
